packages/rule-no-broken-links/src/rule.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * @fileoverview This rule verifies that all links and resources the page
+ * uses are available online. Checks for 404, 410, 500 or 503 status
+ */
+
+import * as URL from 'url';
+import { Category } from 'sonarwhal/dist/src/lib/enums/category';
+import { RuleContext } from 'sonarwhal/dist/src/lib/rule-context';
+import {
+    IRule,
+    ElementFound,
+    RuleMetadata,
+    IAsyncHTMLElement
+} from 'sonarwhal/dist/src/lib/types';
+import { debug as d } from 'sonarwhal/dist/src/lib/utils/debug';
+import { RuleScope } from 'sonarwhal/dist/src/lib/enums/rulescope';
+import { Requester } from 'sonarwhal/dist/src/lib/connectors/utils/requester';
+import { NetworkData } from 'sonarwhal/dist/src/lib/types';
+const debug: debug.IDebugger = d(__filename);
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+export default class NoBrokenLinksRule implements IRule {
+    public static readonly meta: RuleMetadata = {
+        docs: {
+            category: Category.other,
+            description: `Rule to flag broken links in the page`
+        },
+        id: 'no-broken-links',
+        schema: [],
+        scope: RuleScope.site
+    };
+
+    public constructor(context: RuleContext) {
+        const requester = new Requester();
+        const brokenStatusCodes = [404, 410, 500, 503];
+
+        /** Stores the urls and it's response status codes */
+        const fetchedUrls: any[] = [];
+
+        /** Returns an array with all the URLs in the given `srcset` attribute or an empty string if none. */
+        const parseSrcSet = (srcset: string): Array<string> => {
+            if (!srcset) {
+                return [];
+            }
+
+            const urls = srcset
+                .split(',')
+                .map((entry) => {
+                    return entry.trim().split(' ')[0].trim();
+                });
+
+            return urls;
+        };
+
+        /**
+         * The callback to handle rejection returned from the `head` method
+         * When DNS resolution fails, it will be handled here (ex : https://thissitedoesnotexist.com/ )
+         */
+        const handleRejection = (error: any, url: string, element: IAsyncHTMLElement) => {
+            debug(`Error accessing {$absoluteUrl}. ${JSON.stringify(error)}`);
+
+            return context.report(url, element, `Broken link found (404 response)`);
```

Comment:
Yes. The Requester will invoke the error handler ( instead of returning a response with header - response status 404) when DNS resolution fails. I think we should flag those cases as 404 pages .

Reply:
Not sure about this TBH. I think we should have another message because 404 is when the content is missing, not when there's a DNS resolution fail. Maybe something like `Broken link found (domain not found)`.

@alrra thoughts?

Reply:
Switched to `Broken link found (domain not found)`